### DISCUSSION TITLE
Strange validation error: [type=bool_type, input_value=True, input_type=Bool]

### DISCUSSION THREAD

**Discussion by tjadevries** [pydantic/pydantic#10978] [Question] [answered]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I hope that this is the right location for raising my question.

I have an application where I read a toml config file, which results in a dictionary that is used to instantiate nested Pydantic dataclasses.
When the toml file has this line: `set_debug = true`, I get this validation error:

`pydantic_core._pydantic_core.ValidationError: 1 validation error for MyRESTApiConfig
fastapi_config.app_config.set_debug
  Input should be a valid boolean [type=bool_type, input_value=True, input_type=Bool]`

When the toml file has this line: `set_debug = "true"` or `set_debug = 1`, there is no problem.

Unfortunately, when I try to isolate a minimal, reproducible example, the problem also goes away.

What is happening and what can I do to diagnose the cause of the problem?

### Python, Pydantic & OS Version

```Text
pydantic version: 2.10.1
        pydantic-core version: 2.27.1
          pydantic-core build: profile=release pgo=false
                 install path: in my virtual env
               python version: 3.12.7 (tags/v3.12.7:0b05ead, Oct  1 2024, 03:06:41) [MSC v.1941 64 bit (AMD64)]
                     platform: Windows-10-10.0.19045-SP0
             related packages: fastapi-0.115.5 mypy-1.13.0 typing_extensions-4.12.2
                       commit: unknown
```

**Comment by Viicos** [comment]
> which results in a dictionary

Can you show the resulting dictionary? Could you provide the definition of the Pydantic dataclass you are using?

**Comment by tjadevries** [comment]
The resulting dictionary: 

`{'fastapi_config': {'app_config': {'set_debug': True, 'title': 'FastAPI demo REST API', 'openapi_url': '/openapi.json', 'api_version': '/api/v1', 'enviroment': 'development'}, 'uvicorn': {'host': '127.0.0.1', 'port': 6543, 'log_level': 'debug', 'reload': 
False}}}`

Pydantic dataclasses:

```py
@attributes_doc
@dataclass(frozen=True)
class FastAPIAppConfigSection(ConfigSectionBase):
    """Config section for a FastAPI server"""

    set_debug: bool = False
    """Whether or not to run the app in debug mode; defaults to False"""

    title: str = "FastAPI Webapp"
    """The title of the FastAPI app; defaults to 'FastAPI Webapp'"""

    openapi_url: str = "/openapi.json"
    """The URL path where OpenAPI JSON will be available; defaults to '/openapi.json'"""

    api_version: str = "/api/v1"
    """The URL path prefix for API versioning; defaults to '/api/v1'"""

    environment: str = "development"
    """The environment in which the app is running, either 'development', 'staging' or 'production'; defaults to 'development'"""
```

```py
@attributes_doc
@dataclass(frozen=True)
class UvicornConfigSection(ConfigSectionBase):
    """Config section for uvicorn."""

    host: str = "127.0.0.1"
    """IP where the server will be hosted on."""

    port: int = 6543
    """Port wehre the server will be hosted on."""

    log_level: str = "debug"
    """Level of logging for uvicorn."""

    reload: bool = False
    """If the server should automaticallly reload. Should be FALSE for production."""
```

```py
@attributes_doc
@dataclass(frozen=True)
class WebAppConfig(ConfigSectionBase):
    """Main config section for the webapp"""

    secure_headers: SecureHeadersConfigSection = SecureHeadersConfigSection()
    app_config: FastAPIAppConfigSection = FastAPIAppConfigSection()
    cert_config: CertificateConfigSection = CertificateConfigSection()
    uvicorn: UvicornConfigSection = UvicornConfigSection()
```
I did not include the SecureHeadersConfigSection or the CertificateConfigSection, they are not in the toml file.

The base classes are from this open source package: https://github.com/StockwatchDev/application_settings

**Comment by tjadevries** [comment]
The Pydantic error message it shows that `input_value=True`, so this leads me to conclude that there is nothing wrong with the dictionary. It also shows `type=bool_type` and `input_type=Bool`; the latter appears to be the problem. But I cannot find where that comes from.
Other booleans, such as `reload`, don't give a problem.

**Comment by Viicos** [comment]
Unfortunately, I can't reproduce. You will have to find a way to create a MRE (ideally with a repo that can be cloned).

**Comment by EDohmen** [comment]
Just a shot in the dark, but do you use any packages such as `numpy` and the `bool` type is set by one of them?
Numpy has e.g. it's own `bool` class `np.bool_`, which could potentially cause issues with type checks, although being a boolean value and looking exactly the same. Caused me once a headache, but maybe other libs have something similar, causing your `Bool` type to be different.

```bash
>>> import numpy as np
>>> a = True
>>> a
True
>>> type(a)
<class 'bool'>
>>> b = np.abs(7) != 0
>>> b
True
>>> type(b)
<class 'numpy.bool_'>
```

**Reply by Viicos** [reply]
Indeed, if the input type were to be a Python boolean, it wouldn't be capitalized (i.e. `bool`, not `Bool`).

**Reply by tjadevries** [reply]
I am not using numpy. The point being made is exactly what I was trying to explain, rather poorly as it seems. Where is that `Bool` coming from?
I will continue to dig for that, but so far I really have no idea. The problem is occurring reproducibly only for this parameter in the actual project; so far unable to reproduce it myself in any other context :(

**Reply by tjadevries** [reply]
After some more testing: pydantic_core reports `input_type=Bool` if it finds a python boolean. So the capitalized `Bool` does not necessarily point to a type from another package.
I'starting to think that my problem is resulting from an issue with pydantic_core.

**Reply by Viicos** [reply]
https://github.com/pydantic/pydantic/discussions/10978#discussioncomment-11381416 was a good first step in providing a MRE. I've tried installing https://github.com/StockwatchDev/application_settings, redefining your classes in some way and used `tomlkit` to load a TOML file but couldn't reproduce.

If you can manage to provide a set of instructions (i.e. what libraries to install, one or multiple files) to reproduce, I'll be able to take a look. I'm probably missing the part on how you are instantiating the model (which TOML file used, which method called, which TOML library used, etc).

**Comment by bschwedler** [comment]
@Viicos Thank you for taking a look at this!

I put together a minimal reproduction that demonstrates this same failure: https://github.com/bschwedler/pydantic-tomlkit-bool

It is interesting that `ExampleDoc.model_validate(doc)` succeeds but `ExampleDoc(**doc)` fails.

**Reply by bschwedler** [reply]
I can workaround this using a custom field validator, but that is less than ideal.

```python
    @field_validator("value", mode="before")
    @classmethod
    def validate_value(cls, value: bool) -> bool:
        return bool(value)
```

**Reply by Viicos** [reply, marked as answer]
Thanks for the clear repro steps.

While it's unclear why it works with `model_validate`, I see that `tomlkit` is actually defining custom types for everything. For instance, booleans are parsed as [`Bool`](https://github.com/python-poetry/tomlkit/blob/635831f1be9b0e107047e74af8ebecc7c0e4b7bf/tomlkit/items.py#L748) classes (and it does _not_ inherit from the builtin `bool`, meaning Pydantic don't know how to parse it).

`tomlkit` is a library meant to parse TOML files by preserving style, and I assume this is the reason they are using custom classes for every type.

Instead, I would recommend using the stdlib [`tomllib`](https://docs.python.org/3/library/tomllib.html) library or the [`tomli`](https://pypi.org/project/tomli/) PyPI backport.

**Reply by tjadevries** [reply]
Thank you very much for your MRE @bschwedler, I have learned from this. And thanks a lot @Viicos for figuring out what the problem is. I will follow your advice.

**Reply by bschwedler** [reply]
Thanks for the details behind this!  I'll do some more spelunking through `tomlkit` to see if there is a better way to handle this.

Unfortunately, our use case involves dynamically updating TOML files through a CLI, and we want to preserve the styling and comments.  As it stands the above workaround works perfectly for us.

**Reply by bschwedler** [reply]
@tjadevries I just spoke to someone I work with who knows `tomlkit` better.

Apparently you can work around this by calling the [unwrap](https://tomlkit.readthedocs.io/en/latest/api/#tomlkit.toml_document.TOMLDocument.unwrap) method on the TOML document! 🙌🏻 

```python
ExampleDoc(**doc.unwrap())
```

instead of

```python
ExampleDoc(**doc)
```

**Reply by StockwatchDev** [reply]
@bschwedler thank you once more.
